### PR TITLE
3/n inter batch parallelism

### LINKED ISSUE #8316: Add a flavor of `training_step` that allows for expressing inter-batch parallelism
### Motivation

When training a model, the most common approach is to process input batches sequentially (Lightning does this today). This makes sense, because in order to achieve synchronous/deterministic training, the processing of batch[n+1] can’t start until the optimization for batch[n] completes. However, this is only partly true, because it only applies to model weight-dependent logic. We can execute the model weight-independent logic out-of-order while still maintaining numerical equivalence. If the model weight-independent logic takes significant amount of wall time and stresses different system resources compared to the rest of the training step, parallelization can lead to increased throughput.

A trivial example of this is hiding the latency of host-to-device copy of input batches behind computational intensive operations. Without parallization, the execution looks like:
```
batch 0: [HtoD][forward][backward]
batch 1:                          [HtoD][forward][backward]
batch 2:                                                   [HtoD][forward][backward]
```
With parallelization, the latency of HtoD copy can be hidden:
```
batch 0: [HtoD][forward][backward]
batch 1:       [HtoD]             [forward][backward]
batch 2:             [HtoD]                          [forward][backward]
```
A more concrete use case for inter-batch parallelism at FB is to accelerate the model parallel technique used in DLRM (https://github.com/facebookresearch/dlrm). DLRM models consist large amount of embedding layers which are sharded across different GPUs. The forward pass for the embedding layers is roughly as follows:
1. All ranks receive global indices
2. The indices are routed to GPUs that hold the corresponding embedding weights (with alltoall collective)
3. All ranks perform local embedding lookup
4. The embedding lookup results are routed back to senders of the indices (with alltoall collective)

As we can see, step 1 and 2 are (1) model weight-independent (2) expensive (3) stresses communication resources. With inter-batch parallelism, we can hide the latency behind the previous batch’s backward pass.

### Pitch

Support a `training_step` flavor with access to the data iterator, so users can express the aforementioned algorithms.

### LINKED ISSUE #8867: [RFC] Introduce `LightningFetcher` and `BatchProcessor`
## 🚀 Feature

Provide a set of building blocks and APIs for PyTorch Lightning users to manipulate fetching and data based to the the LightningModule `training_step`. 

### Motivation

Currently, several features being developed within Lightning requires a formalisation on how data are being fetched such as Fault Tolerant Training to properly handle `iterators state checkpointing`, `iter-batch-parellism` and refactors within the Loops. 
Furthermore, there is a growing need to have a higher control on the data being pass to the `training_step`.  

Concrete goals for this proposal:

* Provide a composable blocks for users to control the flow of their fetching and data passed to the training_step
* Support fault tolerant training with progress tracking in the case iterator is directly passed the training_step.

API Design proposal.

The `LightningFetcher` would provide `inter-batch-parallism` with pre-fetching=1 for its default behaviour.

Further in the future, the `LightningFetcher` could provide an `auto` mode by default which adapts to the user model / data and auto-tune the fetching using `inter-batch-parallism` to maximise through-put.

The code below is a pseudo-code and likely to be modified.

```py
class BachProcessor:
    def __init__(self, trainer):
        self.trainer = trainer

    @property
    def should_fetch(self):
        return True

    def __call__(self, fetcher, batch):
        return batch

class TBPTTBachProcessor:
    def __init__(self, trainer):
        self.trainer = trainer

    @property
    def should_fetch(self):
        return True

    def __call__(self, fetcher, batch):
        return self.trainer.lightning_module.tbptt_split_batch(batch)


class DataLoaderIterBachProcessor:
    def __init__(self, trainer):
        self.trainer = trainer

    @property
    def should_fetch(self):
        return False

    def __call__(self, fetcher, batch):
        return fetcher.iterator


class LightningFetcher

    def __init__(self, prefetch_batches, batch_processor=...):
        ...

    def setup(self, dataloader, batch_to_device: Callable, profiler: ...):
        ...

def select_batch_processor(trainer, user_batch_processor):

    if user_batch_processor:
        return user_batch_processor

    if "dataloader_iter" in signature(trainer.lightning_module.training_step):
        return DataLoaderIterBachProcessor(...)

    elif trainer.lightning_module.truncated_bptt_steps > 1:
        return TBPTTBachProcessor(...)

    return BachProcessor()


fetcher = LightningFetcher(prefetch_batches=1, batch_processor=BachProcessor())
iterator = iter(fetcher)
batch, is_last, batch_idx = next(fetcher)

# user facing API.
Trainer(prefetch_batches=OPTIONS[0, 1, ..., "auto", LightningFetcher])
```

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Blocked by https://github.com/PyTorchLightning/pytorch-lightning/pull/9047/

This PR adds remove entirely InterBatchProcessor and rely on DataFetcher. 

fixes #8867
fixes #8316

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 1]
```diff
@@ -545,12 +550,16 @@ def _build_kwargs(self, batch: Any, batch_idx: int, opt_idx: int, hiddens: Optio
             the keyword arguments for the training step
         """
         # enable not needing to add opt_idx to training_step
-        step_kwargs = OrderedDict([("batch", batch), ("batch_idx", batch_idx)])
+        step_kwargs = OrderedDict({"batch": batch})
```
Comment (carmocca) [inline pytorch_lightning/loops/batch/training_batch_loop.py:553]: Using list of tuples is important as dicts do not guarantee order in 3.6

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 2]
```diff
@@ -118,29 +112,17 @@ def advance(self, dataloader_iter: Iterator, **kwargs: Any) -> None:
...
-
-            self.batch_progress.increment_processed()
+        batch_idx, (batch, is_last) = next(dataloader_iter)
+
+        # ------------------------------------
+        # TRAINING_STEP + TRAINING_STEP_END
+        # ------------------------------------
+        self.batch_progress.increment_ready()
+
+        with self.trainer.profiler.profile("run_training_batch"):
```
Comment (carmocca) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:122]: The comment is misplaced

```suggestion
        batch_idx, (batch, is_last) = next(dataloader_iter)
        
        self.batch_progress.increment_ready()

        # ------------------------------------
        # TRAINING_STEP + TRAINING_STEP_END
        # ------------------------------------
        with self.trainer.profiler.profile("run_training_batch"):
```

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/loops/batch/training_batch_loop.py:553]: Done !

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py]: Done !

**Comment by tchaton** [comment]
Hey @yifuwang @ananthsub, mind reviewing this PR.

I removed the need for returning `is_last`, added support for accumulate_grad_batches and automatic optimization, optionally passing the `batch_idx`.

Best,
T.C

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 3]
```diff
@@ -250,6 +245,40 @@ def detach_data(model: "pl.LightningModule") -> None:
...
                 loader.unpatch(model)
 
+    def teardown(self) -> None:
+        items = list(vars(self).items())
+        for attr_name, attr in items:
+            if isinstance(attr, AbstractDataFetcher):
+                attr.teardown()
+                delattr(self, attr_name)
+
+    def _check_training_step_requires_dataloader_iter(self) -> bool:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/data_connector.py:255]: don't we want to add this to ConfigValidator?

File: tests/utilities/test_fetching.py [hunk 4]
```diff
@@ -176,6 +180,14 @@ def test_dataloader(self):
 def test_trainer_num_prefetch_batches(tmpdir):
 
     model = RecommenderModel()
+
+    class CheckDataFetcher(Callback):
+        def __init__(self, data_fetcher_cls: Type):
+            self.data_fetcher_cls = data_fetcher_cls
+
+        def on_batch_start(self, trainer, *_) -> None:
+            assert isinstance(trainer.data_connector.train_data_fetcher, self.data_fetcher_cls)
```
Comment (awaelchli) [inline tests/utilities/test_fetching.py:189]: note we are relying on the callback implementation for testing here. this is fragile

**Comment by yifuwang** [comment]
Thanks @tchaton!

> I removed the need for returning is_last, added support for accumulate_grad_batches and automatic optimization, optionally passing the batch_idx.

IIUC when `dataloader_iter` is passed to `training_step`, the contract is:
- `training_step` can prefetch/preprocess arbitrary number of batches from `dataloader_iter`
- However, `training_step` should use and only use the `batch_idx`th batch to optimize the model

Is my understanding correct?

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 5]
```diff
@@ -104,6 +104,11 @@ def run(self, batch: Any, batch_idx: int) -> AttributeDict:
         self.batch_outputs = None  # free memory
         return output
 
+    def on_stop_iteration(self):
+        # if a stop iteration is being triggered within the ``training_step``,
+        # forward the expection to the parent loop.
+        raise StopIteration
```
[the remark this answers is not in the document]
Comment (ananthsub) [reply pytorch_lightning/loops/batch/training_batch_loop.py:110]: +1 - not clear why we need special handling for stop iteration

**Review by ananthsub — COMMENTED** [review]
in addition to @yifuwang 's question, should the `training_step` with `dataloader_iter` be responsible for handling the stop iteration, rather than the loop catching this & doing post-processing before `on_run_end` ? before the contract was that the training step would return the last_batch indicator for the batch processor to handle

File: pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py [hunk 6]
```diff
@@ -199,7 +199,11 @@ def update_eval_epoch_metrics(self) -> _EVALUATE_OUTPUT:
     """
 
     def on_train_split_start(self, batch_idx: int, split_idx: int, split_batch: Any) -> None:
-        self.trainer._results.extract_batch_size(split_batch)
+        # when the user request `dataloader_iter`, we can't track the batch_size
+        # and this is left to user responsability.
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py:203]: Done !

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 7]
```diff
@@ -118,33 +112,22 @@ def advance(self, dataloader_iter: Iterator, **kwargs: Any) -> None:
...
+
+        if not self.trainer.data_connector.train_data_fetcher.store_on_device:
+            with self.trainer.profiler.profile("training_batch_to_device"):
+                batch = self.trainer.accelerator.batch_to_device(batch)
+
+        self.batch_progress.increment_ready()
+
+        # ------------------------------------
+        # TRAINING_STEP + TRAINING_STEP_END
+        # ------------------------------------
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:125]: Removed !

File: pytorch_lightning/trainer/connectors/data_connector.py [hunk 8]
```diff
@@ -250,6 +245,40 @@ def detach_data(model: "pl.LightningModule") -> None:
             if isinstance(loader, _PatchDataLoader):
                 loader.unpatch(model)
 
+    def teardown(self) -> None:
+        items = list(vars(self).items())
+        for attr_name, attr in items:
+            if isinstance(attr, AbstractDataFetcher):
+                attr.teardown()
+                delattr(self, attr_name)
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/data_connector.py:253]: Done !

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/batch/training_batch_loop.py:110]: Here is the following situation.

If a StopIteration is being triggered within the training_step.

On master, this exception is being catched by the TrainingBatchLoop run function which break the while loop.
However, the TrainingEpochLoop isn't stopped and the training continues even if it should have stopped.

Let me refactor this within the on_run_end.

File: pytorch_lightning/trainer/connectors/data_connector.py (hunk 3 above)
Comment (tchaton) [reply pytorch_lightning/trainer/connectors/data_connector.py:255]: Done !

**Comment by tchaton** [comment]
> Thanks @tchaton!
> 
> > I removed the need for returning is_last, added support for accumulate_grad_batches and automatic optimization, optionally passing the batch_idx.
> 
> IIUC when `dataloader_iter` is passed to `training_step`, the contract is:
> 
> * `training_step` can prefetch/preprocess arbitrary number of batches from `dataloader_iter`
> * However, `training_step` should use and only use the `batch_idx`th batch to optimize the model
> 
> Is my understanding correct?

Dear @yifuwang, yes the `training_step` can prefetch/preprocess arbitrary number of batches from `dataloader_iter`.
Internally, Lightning tracks how many batches have been fetched out within the `training_step` and can optionally provide this value to the user through the `batch_idx`. 
This is more for consistency with current API.

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/batch/training_batch_loop.py:110]: I updated based on @awaelchli request, but I prefer the previous solution.

File: pytorch_lightning/loops/epoch/training_epoch_loop.py [hunk 9]
```diff
@@ -81,7 +75,7 @@ def done(self) -> bool:
 
     def connect(
         self,
-        batch_loop: BATCH_LOOP_TYPE = None,
+        batch_loop: TrainingBatchLoop = None,
```
Comment (justusschock) [inline pytorch_lightning/loops/epoch/training_epoch_loop.py:78]:
```suggestion
        batch_loop: Optional[TrainingBatchLoop] = None,
```

**Review by justusschock — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]

File: pytorch_lightning/loops/epoch/training_epoch_loop.py (hunk 9 above)
Comment (tchaton) [reply pytorch_lightning/loops/epoch/training_epoch_loop.py:79]: Done !

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 10]
```diff
@@ -70,6 +70,8 @@ def on_run_start(
         self._dl_max_batches = dl_max_batches
         self._num_dataloaders = num_dataloaders
 
+        self.dataloader_iter = enumerate(dataloader_iter, self.batch_progress.current.ready)
```
Comment (awaelchli) [inline pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:73]: we don't have the attribute set in init

File: tests/core/test_metric_result_integration.py [hunk 11]
```diff
@@ -411,6 +411,7 @@ def training_step(self, batch, batch_idx):
             # However, below we will simulate a failure on `batch_idx=3`.
 
             if self.trainer.fit_loop.restarting:
+
```
Comment (awaelchli) [inline tests/core/test_metric_result_integration.py:414]:
```suggestion
```

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py (hunk 10 above)
Comment (tchaton) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:73]: Done !

File: tests/core/test_metric_result_integration.py (hunk 11 above)
Comment (tchaton) [reply tests/core/test_metric_result_integration.py:414]: Done !

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py (hunk 10 above)
Comment (carmocca) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:74]: Isn't this redundant with `_prepare_dataloader_iter` in the epoch loop?

remove this in favor of the other?

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 12]
```diff
@@ -550,9 +550,13 @@ def _build_kwargs(self, batch: Any, batch_idx: int, opt_idx: int, hiddens: Optio
         step_kwargs = OrderedDict([("batch", batch), ("batch_idx", batch_idx)])
 
         lightning_module = self.trainer.lightning_module
+        training_step_fx = getattr(self.trainer.lightning_module, "training_step")
+
+        if is_param_in_hook_signature(training_step_fx, "dataloader_iter", explicit=True):
+            if not is_param_in_hook_signature(training_step_fx, "batch_idx"):
+                del step_kwargs["batch_idx"]
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/batch/training_batch_loop.py:557]: Trying out ! I believed the test were failing with this :)

File: pytorch_lightning/core/lightning.py [hunk 13]
```diff
@@ -454,6 +454,11 @@ def log(
                     f" of {list(self._metric_attributes.values())}"
                 )
 
+        if is_param_in_hook_signature(self.training_step, "dataloader_iter", explicit=True) and batch_size is None:
+            raise MisconfigurationException(
+                "When the `dataloader_iter` is requested within the `training_step`, `batch_size` should be provided."
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:459]: Done !

File: pytorch_lightning/core/lightning.py [hunk 14]
```diff
@@ -454,6 +454,11 @@ def log(
                     f" of {list(self._metric_attributes.values())}"
                 )
 
+        if is_param_in_hook_signature(self.training_step, "dataloader_iter", explicit=True) and batch_size is None:
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/core/lightning.py:457]: `dataloader_iter` is currently enabled only for training.

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py [hunk 15]
```diff
@@ -70,6 +71,8 @@ def on_run_start(
         self._dl_max_batches = dl_max_batches
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py:71]: Done !

File: pytorch_lightning/core/lightning.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/core/lightning.py:457]: Still, this should be `if self.trainer.training and is_param_in ....`

File: tests/checkpointing/test_trainer_checkpoint.py [hunk 16]
```diff
@@ -75,9 +75,6 @@ def validation_step(self, batch, batch_idx):
         results.append(deepcopy(trainer.callback_metrics))
         best_model_paths.append(trainer.checkpoint_callback.best_model_path)
 
-    for idx in range(len(results) - 1):
-        assert results[idx]["val_loss"] > results[idx + 1]["val_loss"]
-
```
Comment (carmocca) [inline tests/checkpointing/test_trainer_checkpoint.py:80]: Why is this removed? Does it not work anymore?

File: pytorch_lightning/loops/epoch/evaluation_epoch_loop.py (hunk 10 above)
Comment (tchaton) [reply pytorch_lightning/loops/epoch/evaluation_epoch_loop.py]: Updated !

File: tests/utilities/test_fetching.py [hunk 17]
```diff
@@ -189,19 +192,19 @@ def test_trainer_num_prefetch_batches(tmpdir):
         t0 = time()
         trainer = Trainer(**trainer_kwargs)
         trainer.fit(model)
+        assert isinstance(trainer.data_connector.train_data_fetcher, InterBatchParallelDataFetcher)
```
Comment (carmocca) [inline tests/utilities/test_fetching.py:195]: Why move the assertions into the timings?

**Review by carmocca — APPROVED** [review]

File: tests/checkpointing/test_trainer_checkpoint.py (hunk 16 above)
Comment (tchaton) [reply tests/checkpointing/test_trainer_checkpoint.py:80]: No it doesn't.

File: tests/utilities/test_fetching.py (hunk 17 above)
Comment (tchaton) [reply tests/utilities/test_fetching.py:195]: Moved out !

File: tests/utilities/test_fetching.py [hunk 18]
```diff
@@ -236,10 +242,152 @@ def training_step(self, dataloader_iter, batch_idx):
...
+
+    class TestModel(AsyncBoringModel):
+        def __init__(self, trigger_stop_iteration) -> None:
+            super().__init__()
+            self.trigger_stop_iteration = trigger_stop_iteration
+
+        def training_step(self, dataloader_iter: Iterator, *args) -> STEP_OUTPUT:
+            output = super().training_step(dataloader_iter)
+            if self.trigger_stop_iteration and args[0] == EXPECT_NUM_BATCHES_PROCESSED:
+                raise StopIteration
```
Comment (carmocca) [inline tests/utilities/test_fetching.py:330]: @tchaton @awaelchli Is using a `StopIteration` here international or was it just meant as a `DummyException`?

This has the problem that it's the exception checked by the loops to stop, so after it is raised, the loops will try to run `on_run_end` but any further state done at the end of `advance` or `on_advance_end` will not have happened so things can be out of sync.

Blocking #9349

Comment (tchaton) [reply tests/utilities/test_fetching.py:330]: Yes, this is intentional, but it should work as expected.
